plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt

```diff
@@ -36,33 +42,35 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
             domains = listOf(UnitDomain, NullableDomain, CastingDomain, TypeOfDomain, TypeDomain(classes.values.toList()), AnyDomain),
             fields = SpecialFields.all + fields.values.map { it.toViper() },
             functions = SpecialFunctions.all,
-            methods = SpecialMethods.all + methods.values.filter { it.shouldIncludeInProgram }.map { it.viperMethod }.toList(),
+            methods = SpecialMethods.all + methods.values.mapNotNull { it.viperMethod }.toList(),
         )
 
     fun registerForVerification(declaration: FirSimpleFunction) {
-        val embedding = embedFunction(declaration.symbol)
-        embedding.convertBody(this)
+        val signature = embedSignature(declaration.symbol)
+        // Note: it is important that `viperMethod` is only set later, as we need to
+        // place the embedding in the map before processing the body.
+        val embedding = embedUserFunction(declaration.symbol, signature)
+        embedding.viperMethod = convertMethodWithBody(declaration, signature)
     }
 
-    override fun embedFunction(symbol: FirFunctionSymbol<*>): MethodEmbedding {
-        val signature = embedSignature(symbol)
-        return methods.getOrPut(signature.name) {
-            val contractVisitor = ContractDescriptionConversionVisitor(this@ProgramConverter, signature)
-
-            val preconditions = signature.formalArgs.flatMap { it.invariants() } +
-                    signature.formalArgs.flatMap { it.accessInvariants() } +
-                    contractVisitor.getPreconditions(symbol)
-
-            val postconditions = signature.formalArgs.flatMap { it.accessInvariants() } +
-                    signature.params.flatMap { it.dynamicInvariants() } +
-                    signature.returnVar.invariants() +
-                    signature.returnVar.provenInvariants() +
-                    contractVisitor.getPostconditions(symbol)
-
-            UserMethodEmbedding(signature, preconditions, postconditions, symbol)
+    fun embedUserFunction(symbol: FirFunctionSymbol<*>, signature: FullNamedFunctionSignature): UserFunctionEmbedding {
+        when (val existing = methods[signature.name]) {
+            is UserFunctionEmbedding -> return existing
+            else -> {}
```

Comment:
I think you don't need to put an else here

Reply:
it can be also written without creating a local variable `methods[signature.name]?.let { if (it is UserFunctionEmbedding) return it }`